reports/queries.py

```diff
@@ -0,0 +1,288 @@
+from django.conf import settings
+from django.db.models import Count, F, OuterRef, Q, QuerySet, Subquery, Sum
+
+from dining.models import DiningEntry
+from userdetails.models import User, UserMembership
+
+# These queries don't have test cases because they are for reporting and not critical.
+# I have tested them manually and verified the generated SQL queries. The SQL queries
+# are at the bottom.
+
+
+def diner_counts(dining_lists: QuerySet, verified_only=True) -> QuerySet:
+    """Groups by association and annotates the dining lists with entry counts.
+
+    A diner is only counted as member of the association when the membership is
+    verified.
+
+    Returns:
+        A dictionary queryset with the keys `association`, `total_diners`,
+        `association_diners`, `outside_diners` and `guests`.
+    """
+    # The distinct is necessary because the generated join returns a separate row
+    # for each (dining entry, membership) combination. If someone has two
+    # memberships, they will have two rows with the same dining entry.
+    association_count = Count(
+        "dining_entries",
+        filter=Q(
+            dining_entries__external_name="",
+            dining_entries__user__usermembership__association=F("association"),
+            dining_entries__user__usermembership__is_verified=True,
+        )
+        if verified_only
+        else Q(
+            dining_entries__external_name="",
+            dining_entries__user__usermembership__association=F("association"),
+        ),
+        distinct=True,
+    )
+
+    return dining_lists.values("association").annotate(
+        total_diners=Count("dining_entries", distinct=True),
+        association_diners=association_count,
+        # Internal diner count minus association diner count
+        outside_diners=Count(
+            "dining_entries", filter=Q(dining_entries__external_name=""), distinct=True
+        )
+        - association_count,
+        guests=Count(
+            "dining_entries", filter=~Q(dining_entries__external_name=""), distinct=True
+        ),
+    )
+
+
+# SQL for diners count
```

Comment:
The idea of Django is that you don't have to focus on SQL queries. Is the reason you placed the commented SQL queries purely as a replacement for test-cases? If so, honestly, I think it's not neccessary honestly as long as the function is well documented in what it intends to do.

Reply:
I used the generated SQL queries to come up with the correct QuerySet code and check that it's correct. It's much easier to see what is happening in the SQL and whether I would need to use `DISTINCT` somewhere. I left the queries there just for documentation.

The idea for Django's ORM is not to not use SQL, but as an abstraction so that you can use multiple databases with their own SQL dialects, and as a mapper between rows and Python objects ('Object Relational Mapping'). There is nothing wrong with SQL, in fact I would prefer it if the object mapping wasn't such a great feature of Django.